launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java

```diff
@@ -528,13 +581,41 @@ public SparkAppHandle startApplication(SparkAppHandle.Listener... listeners) thr
     pb.environment().put(LauncherProtocol.ENV_LAUNCHER_PORT,
       String.valueOf(LauncherServer.getServerInstance().getPort()));
     pb.environment().put(LauncherProtocol.ENV_LAUNCHER_SECRET, handle.getSecret());
+    pb.environment().put(LauncherProtocol.ENV_LAUNCHER_STOP_FLAG, String.valueOf(stopIfInterrupted));
     try {
       handle.setChildProc(pb.start(), loggerName);
     } catch (IOException ioe) {
       handle.kill();
       throw ioe;
     }
+    return handle;
+  }
+
+  private SparkAppHandle startApplicationAsThread(SparkAppHandle.Listener... listeners) throws IOException {
+    ChildThreadAppHandle handle = LauncherServer.newAppThreadHandle();
+    for (SparkAppHandle.Listener l : listeners) {
+      handle.addListener(l);
+    }
 
+    String appName = getAppName();
+    setConf(LAUNCHER_INTERNAL_PORT,String.valueOf(LauncherServer.getServerInstance().getPort()));
+    setConf(CHILD_PROCESS_LAUNCHER_INTERNAL_SECRET, handle.getSecret());
+    setConf(CHILD_PROCESS_LAUNCHER_STOP_FLAG, String.valueOf(stopIfInterrupted));
+    setConf(CHILD_THREAD_LAUNCHER_ENABLED,"true");
+    try {
+      // It is important that spark-submit class is available in the classpath.
+      // Trying to see if method is available in the classpath else throws Exception.
+      Method main = SparkSubmitRunner.getSparkSubmitMain();
+      Thread submitJobThread = new Thread(new SparkSubmitRunner(main, builder.buildSparkSubmitArgs()));
+      submitJobThread.setName(appName);
+      submitJobThread.setDaemon(true);
+      handle.setChildThread(submitJobThread);
+      submitJobThread.start();
+    } catch (ClassNotFoundException cnfe) {
+      throw new IOException(cnfe);
```

Comment:
minor, but you could provide a better error message here like "make sure spark-core is in your classpath".